File: retrocast-rs/crates/retrocast-core/src/adapters/multistepttl.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

use super::{Adapter, RawRouteEntry, common};
use crate::{
    chem,
    error::Result,
    model::{Molecule, Route, Target},
    route::AdaptMode,
};

#[derive(Clone, Debug, Deserialize, Serialize)]
struct TtlReaction {
    product: String,
    reactants: Vec<String>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
struct TtlRoute {
    reactions: Vec<TtlReaction>,
    #[serde(default)]
    metadata: Map<String, Value>,
}

#[derive(Clone, Copy, Debug, Default)]
pub struct MultiStepTtlAdapter;

impl Adapter for MultiStepTtlAdapter {
    fn name(&self) -> &'static str {
        "multistepttl"
    }

    fn entries(&self, payload: Value, source_key: Option<&str>) -> Result<Vec<RawRouteEntry>> {
        let routes: Vec<TtlRoute> = serde_json::from_value(payload)
            .map_err(|error| common::schema(self.name(), format!("invalid route list: {error}")))?;
        Ok(routes
            .into_iter()
            .enumerate()
            .map(|(index, route)| RawRouteEntry {
                payload: serde_json::to_value(route).expect("TTL route is serializable"),
                source_key: source_key.map(str::to_owned),
                source_row_index: None,
                source_record_id: None,
                target_hint_id: None,
                target_hint_smiles: None,
                source_order: Some(index + 1),
            })
            .collect())
    }

    fn cast(&self, raw_route: Value, mode: AdaptMode, target: Option<&Target>) -> Result<Route> {
        let route: TtlRoute = serde_json::from_value(raw_route)
            .map_err(|error| common::schema(self.name(), format!("invalid route: {error}")))?;
        if route.reactions.is_empty() {
            let target = target.ok_or_else(|| {
                common::logic(
                    self.name(),
                    "adapter.route_transform_failed",
                    "zero-reaction route needs a target",
                )
            })?;
            let (smiles, inchikey) = chem::normalize(&target.smiles)?;
            return Ok(Route {
                target: Molecule {
                    smiles,
                    inchikey,
                    product_of: None,
                    annotations: Map::new(),
                },
                annotations: route.metadata,
                schema_version: Default::default(),
            });
        }

        let root_raw = &route.reactions[0].product;
        let root_smiles = match chem::normalize(root_raw) {
            Ok((smiles, _)) => smiles,
            Err(_) if mode == AdaptMode::Prune => {
                return Err(common::target_pruned(self.name()));
            }
            Err(error) => return Err(error),
        };
        if let Some(target) = target {
            common::require_target_match(&root_smiles, &target.smiles, &target.id, self.name())?;
        }
        let mut precursor_map = BTreeMap::new();
        for reaction in route.reactions {
            match chem::normalize(&reaction.product) {
                Ok((product, _)) => {
                    precursor_map.insert(product, reaction.reactants);
                }
                Err(_) if mode == AdaptMode::Prune => {}
                Err(error) => return Err(error),
            }
        }
        let target_molecule = common::build_from_precursor_map(
            &root_smiles,
            &precursor_map,
            self.name(),
            mode,
            None,
        )?
        .ok_or_else(|| common::target_pruned(self.name()))?;
        Ok(Route {
            target: target_molecule,
            annotations: route.metadata,
            schema_version: Default::default(),
        })
    }
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::MultiStepTtlAdapter;
    use crate::{adapters::Adapter, model::Target, route::AdaptMode};

    fn target() -> Target {
        let (smiles, inchikey) = crate::chem::normalize("CCO").unwrap();
        Target {
            id: "target".to_owned(),
            smiles,
            inchikey,
            acceptable_routes: Vec::new(),
            annotations: Default::default(),
        }
    }

    #[test]
    fn preserves_metadata_and_duplicate_reactants() {
        let route = MultiStepTtlAdapter
            .cast(
                json!({
                    "reactions": [{"product": "CCO", "reactants": ["C", "C"]}],
                    "metadata": {"score": 0.7}
                }),
                AdaptMode::Strict,
                Some(&target()),
            )
            .unwrap();
        assert_eq!(route.leaves().len(), 2);
        assert_eq!(route.annotations["score"], 0.7);
    }

    #[test]
    fn accepts_zero_reaction_route_only_with_target() {
        assert!(
            MultiStepTtlAdapter
                .cast(json!({"reactions": []}), AdaptMode::Strict, Some(&target()))
                .is_ok()
        );
        assert!(
            MultiStepTtlAdapter
                .cast(json!({"reactions": []}), AdaptMode::Strict, None)
                .is_err()
        );
    }
}

File: packages/retrocast-py/src/retrocast/adapters/multistepttl.py
from __future__ import annotations

from collections.abc import Iterator
from typing import Any

from pydantic import BaseModel, Field, RootModel, ValidationError

from retrocast.adapters.base import AdaptMode, RawRouteEntry
from retrocast.adapters.common import build_molecule_from_precursor_map
from retrocast.adapters.errors import adapter_schema_error, adapter_target_mismatch
from retrocast.chem import canonicalize_smiles, get_inchi_key
from retrocast.exceptions import AdapterLogicError, InvalidSmilesError
from retrocast.models.route import Molecule, Route
from retrocast.models.task import Target

# SECTION: Raw MultiStepTTL Schema


class TtlReaction(BaseModel):
    product: str
    reactants: list[str]


class TtlRoute(BaseModel):
    reactions: list[TtlReaction]
    metadata: dict[str, Any] = Field(default_factory=dict)


class TtlRouteList(RootModel[list[TtlRoute]]):
    pass


# SECTION: Adapter


class MultiStepTTLAdapter:
    def iter_raw_routes(self, raw_payload: Any, *, source_key: str | None = None) -> Iterator[RawRouteEntry]:
        target_id = source_key or "<unknown>"
        try:
            routes = TtlRouteList.model_validate(raw_payload)
        except ValidationError as exc:
            raise adapter_schema_error("multistepttl", target_id, "invalid pre-processed route list") from exc
        for source_order, route in enumerate(routes.root, start=1):
            yield RawRouteEntry(payload=route, source_key=source_key, source_order=source_order)

    def cast(self, raw_route: Any, *, mode: AdaptMode = "strict", target: Target | None = None) -> Route:
        target_id = target.id if target is not None else "<unknown>"
        try:
            route = TtlRoute.model_validate(raw_route)
        except ValidationError as exc:
            raise adapter_schema_error("multistepttl", target_id, "invalid route") from exc

        if not route.reactions:
            if target is None:
                raise AdapterLogicError(
                    "MultiStepTTL zero-reaction route needs a target",
                    code="adapter.route_transform_failed",
                    context={"adapter": "multistepttl", "target_id": target_id},
                )
            target_smiles = canonicalize_smiles(target.smiles)
            return Route(
                target=Molecule(smiles=target_smiles, inchikey=get_inchi_key(target_smiles)),
                annotations=route.metadata,
            )

        try:
            root_smiles = canonicalize_smiles(route.reactions[0].product)
        except InvalidSmilesError:
            if mode == "prune":
                raise AdapterLogicError(
                    "MultiStepTTL target molecule was pruned",
                    code="adapter.target_pruned",
                    context={"adapter": "multistepttl", "target_id": target_id},
                ) from None
            raise
        if target is not None:
            expected_smiles = canonicalize_smiles(target.smiles)
            if root_smiles != expected_smiles:
                raise adapter_target_mismatch(
                    "multistepttl", target.id, expected_smiles=expected_smiles, actual_smiles=root_smiles
                )
        precursor_map = {}
        for reaction in route.reactions:
            try:
                product_smiles = canonicalize_smiles(reaction.product)
            except InvalidSmilesError:
                if mode == "strict":
                    raise
                continue
            precursor_map[product_smiles] = reaction.reactants
        route_target = build_molecule_from_precursor_map(root_smiles, precursor_map, adapter="multistepttl", mode=mode)
        if route_target is None:
            raise AdapterLogicError(
                "MultiStepTTL target molecule was pruned",
                code="adapter.target_pruned",
                context={"adapter": "multistepttl", "target_id": target_id},
            )
        return Route(target=route_target, annotations=route.metadata)
